include/irr/asset/IImageAssetHandlerBase.h

```diff
@@ -50,65 +50,93 @@ class IImageAssetHandlerBase : public virtual core::IReferenceCounted
-				region.bufferOffset = copyRegion->bufferOffset;
-				region.bufferRowLength = copyRegion->bufferRowLength * asset::getTexelOrBlockBytesize(EF_R8G8B8_SRGB);
-				region.bufferImageHeight = copyRegion->bufferImageHeight;
-				region.imageOffset = copyRegion->imageOffset;
-				region.imageExtent = copyRegion->imageExtent;
-
-				CMatchedSizeInOutImageFilterCommon::state_type state;
-				state.extent = copyImageParams.extent;
-
-				switch (copyImageParams.format)
-				{
-				case asset::EF_R8_SRGB:
-				{
-					copyImageParams.format = EF_R8G8B8_SRGB;
-					newConvertedImage = ICPUImage::create(std::move(copyImageParams));
-					newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
-					state.inImage = newConvertedImage.get();
-
-					CConvertFormatImageFilter<EF_R8_SRGB, EF_R8G8B8_SRGB> convertFiler;
-					if (!convertFiler.execute(&state))
-						os::Printer::log("LOAD PNG: something went wrong while converting from R8 to R8G8B8 format!", ELL_WARNING);
-				}
-				break;
+				auto referenceImageParams = image->getCreationParameters();
+				auto referenceBuffer = image->getBuffer();
+				auto referenceRegions = image->getRegions();
+
+				const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(outputFormat);
+				auto newImageParams = referenceImageParams;
+				auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceBuffer->getSize() * newTexelOrBlockByteSize);
+				auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(referenceRegions.size());
+
+				newImageParams.format = outputFormat;
 
-				default:
+				uint64_t regionItr = {};
+				for (auto newRegion = newRegions->begin(); newRegion != newRegions->end(); ++newRegion)
 				{
-					newConvertedImage = std::move(copyImageForConverting);
-				}
-				break;
+					auto referenceRegion = referenceRegions.begin() + regionItr++;
+
+					*newRegion = *referenceRegion;
+					newRegion->bufferOffset = referenceRegion->bufferOffset * newTexelOrBlockByteSize;
 				}
+			
+				CONVERSION_FILTER::state_type state;
+				state.extent = newImageParams.extent;
+				state.extentLayerCount = ? ; // why it's vector?
+				state.inBaseLayer = 0;
+				state.inImage = image.get();
+				state.inMipLevel = ? ; // it's for per mip map, so region?
+				state.inOffset = ? ; //old region->bufferOffset ?
+				state.inOffsetBaseLayer = ? ; // probably old single layer byte size
+				state.layerCount = newImageParams.arrayLayers;
+				state.outBaseLayer = 0; // ?
+				state.outMipLevel = ? ; // seems that it's per region
+				state.outOffset = ? ; // new region->bufferOffset * R8G8B8 byte size ?
+				state.outOffsetBaseLayer = ? ; // new offset in bytes to layer ? old layer byte size * R8G8B8 byte size
+
+				newConvertedImage = ICPUImage::create(std::move(newImageParams));
+				newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
```

Comment:
create the image before you execute the filter on it (you will need to execute filter once per mip level)